piControlMain.c

```diff
@@ -1124,7 +1125,64 @@ static long piControlIoctl(struct file *file, unsigned int prg_nr, unsigned long
 			rt_mutex_unlock(&piCore_g.lockUserTel);
 		}
 		break;
+	case KB_AIO_CALIBRATE:
+	{
+		int i;
+		bool found = false;
+		struct pictl_calibrate cali;
+		SMioCalibrationRequest req;
+
+		if (piDev_g.machine_type != REVPI_CORE
+			&& piDev_g.machine_type != REVPI_CONNECT) {
+			return -EPERM;
+		}
+
+		if (!isRunning()) {
+			return -EFAULT;
+		}
 
+		if (copy_from_user(&cali, (const void __user *) usr_addr,
+					sizeof(cali))) {
+			pr_err("failed to copy calibrate request data from user\n");
+			return -EFAULT;
+		}
+
+		for (i = 0; i < RevPiDevice_getDevCnt() && !found; i++) {
+			if (RevPiDevice_getDev(i)->i8uAddress == cali.address
+				&& RevPiDevice_getDev(i)->i8uActive
+				&& RevPiDevice_getDev(i)->sId.i16uModulType
+					== KUNBUS_FW_DESCR_TYP_PI_MIO) {
+				found = true;
+				break;
+			}
+		}
+
+		if (!found || cali.channels == 0) {
+			pr_info("piControlIoctl: calibrate failed,channel 0x%x\n",
+								cali.channels);
+			return -EINVAL;
+		}
+
+		revpi_io_build_header(&req.uHeader, cali.address,
+			sizeof(SMioCalibrationRequestData), IOP_TYP1_CMD_DATA6);
+		req.sData.i8uCalibrationMode = cali.mode;
```

Comment:
Are all modes and channels valid? If not this should be checked here, too, and in case of invalid values an -EINVAL be returned.